### ISSUE TITLE
When to use paket.dependencies and paket.references?

### ISSUE THREAD

**Issue by exercitusvir** [fsprojects/Paket#931]
I've read the documentation on paket.dependencies and paket.references, but it's still not clear to me which one I need to edit. It says that paket.dependencies specifies the rules for the application's dependencis. paket.references adds references to an MSBuild project and apparently it's optional.

But what happens when there are no paket.references? Will Visual Studio still recognize the dependencies? Should I place paket.dependencies in each project folder?

After automatically converting my package.config files with paket, I now have a single paket.dependencies file in the root of my solution and paket.references files in each project folder of my solution. If I want to add dependencies, do I only need to add them to the paket.references files? Will paket.dependencies in the solution folder be then updated automatically?

Basically, I don't understand the possible interaction between paket.dependencies and paket.references.

**Comment by forki** [comment]
paket.dependencies is a global view of your dependencies. there is no such thing in nuget.

paket.references contains all direct dependencies of a csproj or fsproj. It's not optional. You need to specify that in order to get the references in your project.

so in short: you need to edit both, but commands like `paket add` help you there.

**Comment by isaacabraham** [comment]
paket.dependencies = solution wide set of dependencies, not related to any project.
paket.references = dependencies required by a specific fsproj / csproj project etc.

**Comment by exercitusvir** [comment]
Thanks for the quick answers! They really helped clarify their purpose and relationship. You may want to add that to the index page.

I guess I was confused because it explicitly says that it is optional in [The paket.references files](http://fsprojects.github.io/Paket/references-files.html):
> It acts a lot like NuGet's packages.config files but there are some key differences:
...
> - **It's optional**

Is that a mistake or is that meant differently?

**Comment by forki** [comment]
@exercitusvir please send a pull request with whatever you think would clarify the docs. We need the help of newcomers here.

**Comment by isaacabraham** [comment]
It *is* optional insofar as you don't need it just to manage dependencies. I often use just a dependencies file when working with F# scripts when I don't have a project.

You *do* need it if you want to marry dependencies to VS projects (AFAIK).

**Comment by exercitusvir** [comment]
OK. Let me get this working first and then I'll try to improve the documentation for beginners.

One more question. Is there a command to check all paket.references files and update the paket.dependencies file correspondingly? That way, I would only have to edit the paket.references files and run paket with that command just once. 

Another suggestion would be to add an overview of the commands with a short description. I only found the list on the right of the index page where it seems that I have to open each page of a command to find out what it's for.

**Comment by exercitusvir** [comment]
One last question. If I don't have paket.references for each project then how is paket going to know which dependencies a project needs? Will it simply include all dependencies listed in paket.dependencies in every project?

**Comment by forki** [comment]
If you don't have a references file, then your project has 0 references. 

There is no command that puts dependencies from references files into dependencies file. But you will get errors if that happens. 

Yes, a table with the commands would be nice.

**Comment by exercitusvir** [comment]
So that means that paket.references files are in fact required if the project requires dependencies? I just want to make sure I understand it so that I don't put wrong information in the docs.

Are you planning to add a command that puts dependencies from references files into dependencies file?

**Comment by isaacabraham** [comment]
I seem to recall an issue about references -> dependencies file a long time ago. One relatively easy way to add to dependencies to a project is (I think) by using ```paket add```. There's an optional argument on that command which can take in the name of a project. Paket will then get the dependency (if it doesn't already have it) and then add it to the project.

An alternative that I sometimes follow is if it dependency already exists, I manually add the dependency to the references file and then run ```paket install```.

**Comment by forki** [comment]
Yes it's required. We should remove the "optional" remark since it's obviously misleading. 

No I don't plan such a command. Earlier versions did that automatically during install or update. That was convenient, but led to misunderstandings. So we decided to remove that.

**Comment by kflu** [comment]
@isaacabraham -

> An alternative that I sometimes follow is if it dependency already exists, I manually add the dependency to the references file and then run paket install.

It seems paket install would simply no-op if the paket.lock exits and is up-to-date. What works for me is `paket install --force`. Is there a better/more explicit way to tell paket to force update fsproj from `paket.references` without a full reinstallation?

**Comment by isaacabraham** [comment]
Pretty sure that install will update the project files for all references if needed even if lock hasn't changed
